src/connection.rs

```diff
@@ -0,0 +1,205 @@
+            UserState(data) => {
+                // Replace or append user
+                if let Some(index) = self
+                    .users
+                    .iter()
+                    .position(|u| u.get_session() == data.get_session())
+                {
+                    // If there is a new channel id we update it
+                    if data.has_channel_id() {
+                        self.users[index].set_channel_id(data.get_channel_id());
+                    }
+                } else {
+                    // Add user
+                    self.users.push(*data.clone());
+                }
+            }
+            // Ignore other messages
+            _ => {}
+        };
+
+        Ok(message)
+    }
+
+    pub fn send<M: Message + HasType>(&mut self, message: &M) {
+        // Create an stream to ouput this message
+        let mut output_stream = CodedOutputStream::new(&mut self.stream);
+
+        // Create vector of bytes to output
+        let mut result = Vec::new();
+
+        // Append id
+        result.write_u16::<BigEndian>(message.get_id()).unwrap();
+
+        // Create payload from message
+        let mut payload = Vec::new();
+        message.write_to_vec(&mut payload).unwrap();
+
+        // Append length
+        result.write_u32::<BigEndian>(payload.len() as u32).unwrap();
+
+        // Append payload
+        result.append(&mut payload);
+
+        //println!("Sending bytes: {:x?}", result);
+
+        // Send bytes
+        output_stream.write_raw_bytes(&result).unwrap();
+        output_stream.flush().unwrap();
+    }
+
+    pub fn send_version(&mut self) {
+        // Send version message
+        let message = mumble::Version::new();
+        self.send(&message);
+    }
+
+    pub fn send_authentication(&mut self, username: &str, password: &str) {
+        // Send authentication message
+        let mut message = mumble::Authenticate::new();
+        message.set_username(username.to_string());
+        message.set_password(password.to_string());
+        self.send(&message);
+    }
+
+    pub fn send_text_message(&mut self, text: &str) {
+        // Send text message
+        match self.session {
+            Some(s) => {
+                let mut message = mumble::TextMessage::new();
+                message.set_actor(s);
+                message.set_channel_id(vec![0]);
+                message.set_message(text.to_string());
+                self.send(&message);
+            }
+            None => println!("Connection not established. (No session)"),
+        }
+    }
+
+    pub fn change_channel(&mut self, channel_id: u32) {
+        // Send channel change message
+        if let Some(s) = self.session {
+            let mut message = mumble::UserState::new();
+            message.set_session(s);
+            message.set_actor(s);
+            message.set_channel_id(channel_id);
+            self.send(&message)
+        } else {
+            println!("Connection not established. (No session)");
```

Comment:
Instead of using `println!`, please add a `log` crate and use the `debug!` macro. Or at least open an issue for that.